src/test/java/jnr/unixsocket/SocketInteropTest.java

```diff
@@ -155,4 +171,42 @@ public void run() {
             }
         }.start();
     }
+
+    @Test
+    public void concurrentReadAndWrite() throws IOException {
+        // https://bugs.openjdk.java.net/browse/JDK-4774871
+        assumeTrue("the TCP channels API doesn't support concurrent read and write",
+                socketPairFactory != TcpChannelsApiSocketPair.FACTORY);
+
+        socketPair.connectBlocking();
+
+        // Helper thread sleeps 500 ms, then writes two messages.
+        new Thread(getClass().getName() + ".concurrentReadAndWrite") {
+            @Override
+            public void run() {
+                try {
+                    Thread.sleep(500);
```

Comment:
why do we need to sleep here? Would a latch be useful instead?

Reply:
I’d love to use a latch but I want to signal it _after_ the stream read starts but before it ends. Otherwise it’s still racy.